### PR TITLE
[SYCL] Allow gcc asm statements in kernel code.

### PR DESCRIPTION
**Pull request by premanandrao**
Signed-off-by: Premanand M Rao <premanand.m.rao@intel.com>

### CODE REVIEW

**Comment by premanandrao** [comment]
> This likely needs a codegen test to show what actually happens. Have you done an end-to-end test to show that this still works?

I have run them by hand (only), but there is someone further downstream who has a whole bunch of these tests and have validated the results.

**Comment by premanandrao** [comment]
> This likely needs a codegen test to show what actually happens.

I added a test recreated from a supplied example to check the asm part.

File: clang/test/CodeGenSYCL/inline_asm.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    class kernel {};
+  }
+}
+
+using namespace cl::sycl;
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  // CHECK: %[[ARRAY_A:[0-9a-z]+]] = alloca [100 x i32], align 4
+  // CHECK-NEXT: %0 = bitcast [100 x i32]* %[[ARRAY_A]] to i8*
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/inline_asm.cpp]: I don't think lifetime markers are worth testing.

Comment (erichkeane) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Also, these both use %0, which might change depending on naming conventions.  Those shouldn't be in the test.

Comment (premanandrao) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Thanks, fixed.

File: clang/test/CodeGenSYCL/inline_asm.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    class kernel {};
+  } // sycl
+} // cl
+
+using namespace cl::sycl;
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  // CHECK: %[[ARRAY_A:[0-9a-z]+]] = alloca [100 x i32], align 4
+  // CHECK-NEXT: %[[NUM:[0-9]+]] = bitcast [100 x i32]* %[[ARRAY_A]] to i8*
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/inline_asm.cpp]: This is only there for the lifetime marker, so no need for htis check line either.

Comment (premanandrao) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Thanks again, Erich!

**Comment by keryell** [comment]
Interesting extension!

File: clang/test/CodeGenSYCL/inline_asm.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-linux-sycldevice -S -emit-llvm -x c++ %s -o - | FileCheck %s
+
+inline namespace cl {
+namespace sycl {
+class kernel {};
+} // namespace sycl
+} // namespace cl
```
Comment (bader) [inline clang/test/CodeGenSYCL/inline_asm.cpp]:
```suggestion
class kernel;
```

File: clang/test/CodeGenSYCL/inline_asm.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-linux-sycldevice -S -emit-llvm -x c++ %s -o - | FileCheck %s
```
Comment (bader) [inline clang/test/CodeGenSYCL/inline_asm.cpp]:
```suggestion
// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -emit-llvm -x c++ %s -o - | FileCheck %s
```

File: clang/test/CodeGenSYCL/inline_asm.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+inline namespace cl {
+namespace sycl {
+class kernel {};
+} // namespace sycl
+} // namespace cl
+
+using namespace cl::sycl;
```
Comment (bader) [inline clang/test/CodeGenSYCL/inline_asm.cpp]:
```suggestion

```

File: clang/test/CodeGenSYCL/inline_asm.cpp (hunk 3 above)
Comment (premanandrao) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Thanks, done.

File: clang/test/CodeGenSYCL/inline_asm.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Done.

File: clang/test/CodeGenSYCL/inline_asm.cpp (hunk 5 above)
Comment (premanandrao) [reply clang/test/CodeGenSYCL/inline_asm.cpp]: Done.

**Comment by premanandrao** [comment]
@bader, if you get a chance, please merge this.   Thank you!

**Pull request merged**